overlord/ifacestate/hotplug.go

```diff
@@ -30,3 +34,115 @@ func (m *InterfaceManager) HotplugDeviceAdded(devinfo *hotplug.HotplugDeviceInfo
 }
+
+// ensureUniqueName modifies proposedName so that it's unique according to isUnique predicate.
+// Uniqueness is achieved by appending a numeric suffix, or increasing existing suffix.
+func ensureUniqueName(proposedName string, isUnique func(string) bool) string {
+	// if the name is unique right away, do nothing
+	if isUnique(proposedName) {
+		return proposedName
+	}
+
+	// extract number suffix if present
+	pname := []rune(proposedName)
+	end := len(pname) - 1
+	suffixIndex := end
+	for suffixIndex >= 0 && unicode.IsDigit(pname[suffixIndex]) {
+		suffixIndex--
+	}
+
+	var suffixNumValue uint64
+	// if numeric suffix hasn't been found, append "-" before the number
+	if suffixIndex == end {
+		pname = append(pname, '-')
+		suffixIndex++
+	} else {
+		var err error
+		suffixNumValue, err = strconv.ParseUint(string(pname[suffixIndex+1:]), 10, 32)
+		if err != nil {
+			suffixIndex = end
+		}
+	}
+
+	// increase suffix value until we have a unique name
+	for {
+		suffixNumValue++
+		proposedName = fmt.Sprintf("%s%d", string(pname[0:suffixIndex+1]), suffixNumValue)
+		if isUnique(proposedName) {
+			return proposedName
+		}
+	}
+}
+
+const maxLen = 20
+
+// cleanupSlotName sanitizes proposedName to make it a valid slot name that
+// passes validation rules implemented by ValidateSlotName (see snap/validate.go):
+// - only lowercase letter, digits and dashes are allowed
+// - must start with a letter
+// - no double dashes, cannot end with a dash.
+// In addition names are truncated not to exceed maxLen characters.
+func cleanupSlotName(proposedName string) string {
+	var out []rune
+	var charCount int
+	// the dash flag is used to prevent consecutive dashes, and the dash in the front
+	dash := true
+Loop:
+	for _, c := range proposedName {
+		switch {
+		case c == '-' && !dash:
+			dash = true
+			out = append(out, '-')
+		case unicode.IsLetter(c):
+			out = append(out, unicode.ToLower(c))
+			dash = false
+			charCount++
+			if charCount >= maxLen {
+				break Loop
+			}
+		case unicode.IsDigit(c) && charCount > 0:
+			out = append(out, c)
+			dash = false
+			charCount++
+			if charCount >= maxLen {
+				break Loop
+			}
+		default:
+			// any other character is ignored
+		}
+	}
+	// make sure the name doesn't end with a dash
+	if len(out) > 0 && out[len(out)-1] == '-' {
+		out = out[0 : len(out)-1]
+	}
+	return string(out)
+}
+
+var nameAttrs = []string{"NAME", "ID_MODEL_FROM_DATABASE", "ID_MODEL"}
+
+// suggestedSlotName returns the shortest name derived from attributes defined by nameAttrs, or
+// the fallbackName if there is no known attribute to derive name from. The name created from
+// attributes is cleaned up by cleanupSlotName function.
+// The fallbackName is typically the name of the interface.
+func suggestedSlotName(devinfo *hotplug.HotplugDeviceInfo, fallbackName string) string {
+	var candidates []string
+	for _, attr := range nameAttrs {
+		val, ok := devinfo.Attribute(attr)
```

Comment:
While scanning this I was confused by the use of `val` and `name` below. I think using a single variable name would be easier to follow, at least for me.